exchanges/graphcache/src/operations/shared.ts

```diff
@@ -203,10 +213,15 @@ export const makeSelectionIterator = (
             if (isMatching) {
               if (process.env.NODE_ENV !== 'production')
                 pushDebugNode(typename, fragment);
+
+              const isFragmentOptional = isOptional(select);
               child = makeSelectionIterator(
                 typename,
                 entityKey,
                 defer || isDeferred(select, ctx.variables),
+                isFragmentOptional === undefined
+                  ? optional
+                  : isFragmentOptional,
```

Comment:
I think this _might_ potentially be incorrect.

We pass `optionalRef` back into `optional` as an argument here, i.e.
`optionalRef = optional`, as per the above.

This means that _any sub-selection_ of an optional selection remains optional. In other words, these two cases should be equivalent:

```graphql
query {
  optionalField @_optional {
    nonOptionalField
  }
}

query {
  ... @_optional {
    optionalField {
      nonOptionalField
    }
  }
}
```

However, they're not, as far as I can tell.
We inherit `optionalRef` through sub-iterations, which is fine, since those resolve fragments, however, passing in `optional` as an argument to `optionalRef` seems to break assumptions here.

I previously didn't notice this. Sorry!